### ISSUE TITLE
efficiency-usability problem with PointerHashMapSet iterators

### ISSUE THREAD

**Issue by RiccardoRossi** [KratosMultiphysics/Kratos#2198]
@pooyan-dadvand i believe we have a usability problem which is also an efficiency issue with the iterators of the PointerHashMap.

i uncovered the issue when attempting to make use of std::unique_ptr instead of shared_ptr in the map.
when iterating "normally" over the map, the following line would not compile

      pointer_type operator->() const { return map_iterator->second; }

the cause is that it attempts to copy construct the pointer which is a forbidden operation when using unique_ptr.
my point is that aside for the current map not able to hold unique_ptrs, this behaviour has also a performance impact when iterating.

**Comment by loumalouomega** [comment]
@RiccardoRossi I added the SharedPointerComparator and the SharedPointerHasher in the file key_hash.h, I can add the corresponding for a unique_ptr. I recommend to unify all the hashers and comparors in this file in order to reduce code duplication, or we will have hundreds of comparators and hashers incompatible between them spread among the applications

https://github.com/KratosMultiphysics/Kratos/blob/68cd91f4180fa872e10d932e046a22fe64c73584/kratos/includes/key_hash.h#L152

https://github.com/KratosMultiphysics/Kratos/blob/68cd91f4180fa872e10d932e046a22fe64c73584/kratos/includes/key_hash.h#L170

**Comment by pooyan-dadvand** [comment]
The short answer is Yes this was intentionally. 

In general is not a good practice to pass references of the shared_ptr because you want to have different pointer sharing the same object. 

In this case the reason is the lifetime.  I cannot tie the lifetime of the pointed object to the lifetime of the map. If you return the reference to the pointer and then remove the entity from the map it will crush. I imagine that you have use cases that the owner is the map and for that you want to use the unique_ptr but in general there are real shared cases to be considered. 

About the performance, pointer has a move constructor. So hopefully it will eliminate the overhead of temporary pointers. Nevertheless without profiling I cannot tell the real effect.

**Comment by pooyan-dadvand** [comment]
It seems that there is a semantic error there. Let chek it together

**Comment by philbucher** [comment]
is this still active?

**Comment by pooyan-dadvand** [comment]
@RiccardoRossi?

**Comment by RiccardoRossi** [comment]
yes, this is not solved... i think it would be useful to be able to store unique pointers...
